src/util/tev_curl.c

```diff
@@ -549,385 +546,492 @@ struct tcurl_ctx *tcurl_init(TALLOC_CTX *mem_ctx,
+
+    ret = tcurl_set_option(tcurl_req, CURLOPT_HTTPHEADER, tcurl_req->headers);
+    if (ret != EOK) {
+        goto done;
+    }
+
+    switch (method) {
+    case TCURL_HTTP_GET:
+        /* Nothing to do here. GET is default. */
+        break;
+    case TCURL_HTTP_PUT:
+        ret = tcurl_set_option(tcurl_req, CURLOPT_UPLOAD, 1L);
         if (ret != EOK) {
-            DEBUG(SSSDBG_CRIT_FAILURE,
-                  "Failed to set write callbacks [%d]: %s\n",
-                  ret, sss_strerror(ret));
-            return ret;
+            goto done;
         }
         break;
-    case TCURL_HTTP_GET:
-        /* GET just needs the write callbacks, nothing to do here.. */
+    case TCURL_HTTP_POST:
+        ret = tcurl_set_option(tcurl_req, CURLOPT_POST, 1L);
+        if (ret != EOK) {
+            goto done;
+        }
         break;
     case TCURL_HTTP_DELETE:
-        crv = curl_easy_setopt(state->http_handle,
-                               CURLOPT_CUSTOMREQUEST,
-                               "DELETE");
-        if (crv != CURLE_OK) {
-            DEBUG(SSSDBG_OP_FAILURE,
-                  "Failed to set the uplodad option [%d]: %s\n",
-                  crv, curl_easy_strerror(crv));
-            return EIO;
+        ret = tcurl_set_option(tcurl_req, CURLOPT_CUSTOMREQUEST, "DELETE");
+        if (ret != EOK) {
+            goto done;
         }
         break;
-    default:
-        return EFAULT;
     }
 
-    return EOK;
-}
+    ret = EOK;
 
-static size_t tcurl_http_write_data(char *ptr,
-                                    size_t size,
-                                    size_t nmemb,
-                                    void *userdata)
-{
-    errno_t ret;
-    size_t realsize = size * nmemb;
-    struct sss_iobuf *outbuf = talloc_get_type(userdata, struct sss_iobuf);
-
-    DEBUG(SSSDBG_TRACE_INTERNAL, "---> begin libcurl data\n");
-    DEBUG(SSSDBG_TRACE_INTERNAL, "%s\n", ptr);
-    DEBUG(SSSDBG_TRACE_INTERNAL, "<--- end libcurl data\n");
-
-    ret = sss_iobuf_write_len(outbuf, (uint8_t *) ptr, realsize);
+done:
     if (ret != EOK) {
-        DEBUG(SSSDBG_CRIT_FAILURE,
-              "Failed to write data to buffer [%d]: %s\n", ret, sss_strerror(ret));
-        /* zero signifies an EOF */
-        return 0;
+        talloc_free(tcurl_req);
+        return NULL;
     }
 
-    return realsize;
+    return tcurl_req;
 }
 
-static size_t tcurl_http_read_data(void *ptr,
-                                   size_t size,
-                                   size_t nmemb,
-                                   void *userdata)
+errno_t tcurl_req_enable_rawoutput(struct tcurl_request *tcurl_req)
+{
+    return tcurl_set_option(tcurl_req, CURLOPT_HEADER, 1L);
+}
+
+errno_t tcurl_req_verify_peer(struct tcurl_request *tcurl_req,
+                              const char *capath,
+                              const char *cacert,
+                              bool verify_peer,
+                              bool verify_host)
```

Comment:
I find it hard to read function calls with multiple boolean attributes, wouldn't it be nicer to have flags so that `tcurl_req_verify_peer(req, capath, cacert, true, true)` becomes `tcurl_req_verify_peer(req, capath, cacert, TCURL_REQ_VERIFY_PEER | TCURL_REQ_VERIFY_HOST)